### ISSUE TITLE
How to specify configs for pretrained-torchvision models vs. pretrained-caffe2?

### ISSUE THREAD

**Issue by andrewgrahambrown** [facebookresearch/detectron2#778]
QUESTION:

I wish to know how to specify the configs for fine tuning weights using the pretrained models from torchvision. I believe the issue arises due to the different input normalisations used by default in caffe2 and torchvision. This issue is closely related #752, #450,  #170 and #118, and the 'write-models' section of the guidelines, though I cannot find the actual config parameters needed for this use case in these issues/guidelines.

USE CASE: 

I want to use weights from a torchvision model [e.g. ResNet-50](https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py) as a backbone toreproduce approximately the same results as with detectron2's pretrained models. The ultimate reason is that I want to be sure I can reliably introduce my own pretrained models into detectron2.

So far I have copied the torchvision code into detectron, with the following modifications: (1) the forward pass should output a dictionary to be compatible with detectron2 (outputting 'res2', 'res3' etc) (2) BatchNorms should be converted to frozen batchnorms everywhere (3) Parameters in the stem and the first stage should be frozen (p.requires_grad=False). 

I then want to finetune on COCO and compare with the ResNet50 model that comes with detectron2. I have tried but cannot get closer than 1.5%ish (33ish->31.5ish) AP for COCO/InstanceSegmentation (with hard-coded attempts to handle the caffe2/pytorch input differences). Other than the stride positioning in downsample blocks, the two versions should be the same, and so reach ~the same accuracy, but I cannot get them to do so. I believe this is due to input specifications, but cannot tell exactly.

FURTHER DETAILS: 

By default, it seems detectron2 loads weights from a model that was pretrained in caffe2 (EDIT: This is the original resnet weights from He et al, apparently from caffe (original)). However, I understand caffe2 had different input normalisations / data formats to the Pytorch torchvision models. This implies different configs should be specified, but exactly what these should be I cannot see. The main differences I know / think are: 

INPUT.FORMAT, which is I believe 'BGR' for caffe2 pretraining but 'RGB' for pytorch

PIXEL_MEAN, default setting is [103.53, 116.28, 123.675] (BGR) in detectron2. I guess this corresponds to [0.485, 0.456, 0.406] (RGB) in official pytorch examples [(see line 202)](https://github.com/pytorch/examples/blob/master/imagenet/main.py) but a trivial swapping of these values in the config is not the solution.

PIXEL_STD, which is [1.0, 1.0, 1.0] (BGR) in detectron2 but again I guess corresponds to [0.229, 0.224, 0.225] (RGB) in pytorch examples [(see line 203)](https://github.com/pytorch/examples/blob/master/imagenet/main.py). Again a trivial swapping of these values in the config is not the solution.

FEATURE REQUEST:

Humbly, I would request a feature: a demo where loading pretrained weights from a single torchvision model with the appropriate config is shown (e.g. ResNet-50), and then fine tuning performed on COCO. I realise this request is non-zero work, due to naming conversion, the different class structure in detectron2 (e.g. Conv2D), and the different stride by default (detectron2 has it in the first conv of a downsampling block by default due to caffe2, torchvision in the second).

**Comment by ppwwyyxx** [comment]
> I then want to finetune on COCO and compare with the ResNet50 model that comes with detectron2. I have tried but cannot get closer than 1.5%ish (33ish->31.5ish)

We did observe in the past that torchvision models do not perform as well as our classic models in fine-tuning. They are two models trained with different settings so there should be a gap.  Though the gap shouldn't be as large (should be < 0.5). [mmdet](https://github.com/open-mmlab/mmdetection/blob/master/docs/MODEL_ZOO.md) also reports a gap.

The pre-trained models we use come not from caffe2, but from the official ResNet release in https://github.com/KaimingHe/deep-residual-networks.

Input images to our model will be in range [0, 255] with format `INPUT.FORMAT` (RGB for torchvision models). They will then be normalized by PIXEL_MEAN/STD. So I think PIXEL_MEAN should be `[0.485, 0.456, 0.406] * 255` (or 256 depend on how torchvision normalizes the input), similar for PIXEL_STD. Though these need to be further verified.

**Comment by andrewgrahambrown** [comment]
Thanks for the speedy response!

> The pre-trained models we use come not from caffe2, but from the official ResNet release in https://github.com/KaimingHe/deep-residual-networks.

Ah I see! Interesting to know. The repo seems to indicate that He et al's models were trained in caffe (not caffe2 as I thought). 

> We did observe in the past that torchvision models do not perform as well as our classic models in fine-tuning. They are two models trained with different settings so there should be a gap. Though the gap shouldn't be as large (should be < 0.5). mmdet also reports a gap.

Yes I would also expect this sort of ~0.5% difference. It is extremely difficult to reproduce pretraining/finetuning processes exactly My gap is a bit too large though - the main reason I suspect the input config is nn.SyncBatchNorm gives me +0.5% vs FrozenBatchNorm. If that was supposed to happen, nn.SyncBatchNorm would have been the default setting - thus I expect my FrozenBatchNorms are (effectively) frozen at the wrong value.

>Input images to our model will be in range [0, 255] with format INPUT.FORMAT (RGB for torchvision models). They will then be normalized by PIXEL_MEAN/STD. So I think PIXEL_MEAN should be [0.485, 0.456, 0.406] * 255 (or 256 depend on how torchvision normalizes the input), similar for PIXEL_STD.

I can try this.

**Comment by ppwwyyxx** [comment]
Instructions to use torchvision resnets are provided in 32570b767e1b69516c58734fe6cc46005bab2aae
